File: tests/fixtures/cpp/sample_complex.cpp
#include <vector>
#include <string>
#include <memory>
#include <algorithm>
#include <stdexcept>

namespace geometry {

struct Point {
    double x;
    double y;

    Point(double x, double y) : x(x), y(y) {}

    double distanceTo(const Point& other) const {
        double dx = other.x - x;
        double dy = other.y - y;
        return std::sqrt(dx * dx + dy * dy);
    }
};

class Shape {
public:
    virtual ~Shape() = default;
    virtual double area() const = 0;
    virtual double perimeter() const = 0;
    virtual std::string name() const = 0;
};

class Circle : public Shape {
private:
    Point center_;
    double radius_;
public:
    Circle(Point center, double radius) : center_(center), radius_(radius) {
        if (radius <= 0) {
            throw std::invalid_argument("Radius must be positive");
        }
    }

    double area() const override {
        return 3.14159265358979 * radius_ * radius_;
    }

    double perimeter() const override {
        return 2.0 * 3.14159265358979 * radius_;
    }

    std::string name() const override { return "Circle"; }

    Point getCenter() const { return center_; }
    double getRadius() const { return radius_; }
};

class Rectangle : public Shape {
private:
    Point topLeft_;
    double width_;
    double height_;
public:
    Rectangle(Point topLeft, double width, double height)
        : topLeft_(topLeft), width_(width), height_(height) {
        if (width <= 0 || height <= 0) {
            throw std::invalid_argument("Dimensions must be positive");
        }
    }

    double area() const override {
        return width_ * height_;
    }

    double perimeter() const override {
        return 2.0 * (width_ + height_);
    }

    std::string name() const override { return "Rectangle"; }

    double getWidth() const { return width_; }
    double getHeight() const { return height_; }
};

class ShapeCollection {
private:
    std::vector<std::unique_ptr<Shape>> shapes_;
public:
    void addShape(std::unique_ptr<Shape> shape) {
        shapes_.push_back(std::move(shape));
    }

    double totalArea() const {
        double total = 0.0;
        for (const auto& shape : shapes_) {
            total += shape->area();
        }
        return total;
    }

    double totalPerimeter() const {
        double total = 0.0;
        for (const auto& shape : shapes_) {
            total += shape->perimeter();
        }
        return total;
    }

    size_t count() const { return shapes_.size(); }

    std::vector<std::string> names() const {
        std::vector<std::string> result;
        result.reserve(shapes_.size());
        for (const auto& shape : shapes_) {
            result.push_back(shape->name());
        }
        return result;
    }
};

} // namespace geometry

int main() {
    using namespace geometry;

    ShapeCollection collection;

    auto circle = std::make_unique<Circle>(Point(0.0, 0.0), 5.0);
    collection.addShape(std::move(circle));

    auto rect = std::make_unique<Rectangle>(Point(1.0, 1.0), 4.0, 6.0);
    collection.addShape(std::move(rect));

    std::cout << "Total shapes: " << collection.count() << std::endl;
    std::cout << "Total area: " << collection.totalArea() << std::endl;
    std::cout << "Total perimeter: " << collection.totalPerimeter() << std::endl;

    return 0;
}

File: tests/fixtures/rust/sample_complex.rs
use std::fmt;

#[derive(Debug, Clone)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn distance_to(&self, other: &Point) -> f64 {
        let dx = other.x - self.x;
        let dy = other.y - self.y;
        (dx * dx + dy * dy).sqrt()
    }
}

impl fmt::Display for Point {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {})", self.x, self.y)
    }
}

trait Shape: fmt::Display {
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;
    fn name(&self) -> &str;
}

struct Circle {
    center: Point,
    radius: f64,
}

impl Circle {
    fn new(center: Point, radius: f64) -> Result<Self, String> {
        if radius <= 0.0 {
            return Err("Radius must be positive".to_string());
        }
        Ok(Circle { center, radius })
    }
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }

    fn perimeter(&self) -> f64 {
        2.0 * std::f64::consts::PI * self.radius
    }

    fn name(&self) -> &str {
        "Circle"
    }
}

impl fmt::Display for Circle {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Circle(center={}, radius={})", self.center, self.radius)
    }
}

struct Rectangle {
    top_left: Point,
    width: f64,
    height: f64,
}

impl Rectangle {
    fn new(top_left: Point, width: f64, height: f64) -> Result<Self, String> {
        if width <= 0.0 || height <= 0.0 {
            return Err("Dimensions must be positive".to_string());
        }
        Ok(Rectangle { top_left, width, height })
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn perimeter(&self) -> f64 {
        2.0 * (self.width + self.height)
    }

    fn name(&self) -> &str {
        "Rectangle"
    }
}

impl fmt::Display for Rectangle {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Rectangle(top_left={}, {}x{})", self.top_left, self.width, self.height)
    }
}

struct ShapeCollection {
    shapes: Vec<Box<dyn Shape>>,
}

impl ShapeCollection {
    fn new() -> Self {
        ShapeCollection { shapes: Vec::new() }
    }

    fn add(&mut self, shape: Box<dyn Shape>) {
        self.shapes.push(shape);
    }

    fn total_area(&self) -> f64 {
        self.shapes.iter().map(|s| s.area()).sum()
    }

    fn total_perimeter(&self) -> f64 {
        self.shapes.iter().map(|s| s.perimeter()).sum()
    }

    fn count(&self) -> usize {
        self.shapes.len()
    }
}

fn main() {
    let mut collection = ShapeCollection::new();

    let circle = Circle::new(Point::new(0.0, 0.0), 5.0).unwrap();
    collection.add(Box::new(circle));

    let rect = Rectangle::new(Point::new(1.0, 1.0), 4.0, 6.0).unwrap();
    collection.add(Box::new(rect));

    println!("Total shapes: {}", collection.count());
    println!("Total area: {:.2}", collection.total_area());
    println!("Total perimeter: {:.2}", collection.total_perimeter());
}
